Explorer/Assets/Scripts/ECS/SceneLifeCycle/Systems/GatherGltfAssetsSystem.cs

```diff
@@ -0,0 +1,132 @@
+            entitiesUnderObservation = HashSetPool<EntityReference>.Get();
+        }
+
+        public override void Dispose()
+        {
+            HashSetPool<EntityReference>.Release(entitiesUnderObservation);
+            entitiesUnderObservation = null;
+        }
+
+        protected override void Update(float t)
+        {
+            if (framesLeft > 0)
+            {
+                GatherEntitiesQuery(World);
+
+                framesLeft--;
+            }
+            else if (!concluded)
+            {
+                if (totalAssetsToResolve == -1)
+                    totalAssetsToResolve = entitiesUnderObservation!.Count;
+
+                if (reports == null && !readinessReportQueue.TryDequeue(sceneData.Parcels, out reports))
+                {
+                    // if there is no report to dequeue, nothing to do
+                    concluded = true;
+                    return;
+                }
+
+                concluded = true;
+
+                List<EntityReference> toDelete = ListPool<EntityReference>.Get();
+
+                // iterate over entities
+
+                foreach (EntityReference entityRef in entitiesUnderObservation!)
+                {
+                    // if entity has died
+                    // or entity no longer contains GltfContainerComponent
+                    // continue
+                    if (!entityRef.IsAlive(World)
+                        || !World.TryGet(entityRef, out GltfContainerComponent gltfContainerComponent))
+                    {
+                        toDelete.Add(entityRef);
+                        continue;
+                    }
+
+                    // if Gltf Container Component has finished loading at least once (it can be reconfigured, we don't care)
+                    if (gltfContainerComponent.State.Value == LoadingState.Loading)
+                    {
+                        concluded = false;
+
+                        // no reason to iterate further
+                        break;
+                    }
+
+                    // remove entity from list - it's loaded, we don't need to check it anymore
+                    toDelete.Add(entityRef);
+                }
+
+                assetsResolved += toDelete.Count;
+                float progress = totalAssetsToResolve != 0 ? assetsResolved / (float)totalAssetsToResolve : 1;
+
+                for (var i = 0; i < reports.Count; i++)
+                {
+                    AsyncLoadProcessReport report = reports[i];
+                    report.ProgressCounter.Value = progress;
+                }
+
+                entitiesUnderObservation.ExceptWith(toDelete);
+                ListPool<EntityReference>.Release(toDelete);
+
+                if (concluded)
+                {
+                    for (var i = 0; i < reports.Count; i++)
+                        reports[i].CompletionSource.TrySetResult();
+
+                    // TODO return reports to the pool as they are no longer needed
```

Comment:
Please implement this

Reply:
Solved here: https://github.com/decentraland/unity-explorer/pull/258/commits/13e7890c848e728fd197182e04eeb4e2e166e903 what do you think?